lib/tld_sync.php

```diff
@@ -0,0 +1,135 @@
+
+            foreach ($tlds_pricing as $tld => $pricing) {
+                $this->DomainsTlds->updatePricings(
+                    $tld,
+                    $this->formatPricing($pricing, $tld, $company_id),
+                    $company_id
+                );
+            }
+        }
+    }
+
+    /**
+     * Format a pricing record returned by the registrar module
+     *
+     * @param array $pricing TLDs pricing
+     *    [currency => [year# => ['register' => price, 'transfer' => price, 'renew' => price]]]
+     * @param string $tld The TLD to sync registrar prices
+     * @param int $company_id The ID of the company where the TLDs will be synchronized (optional)
+     * @return array The formatted pricing record
+     */
+    private function formatPricing($pricing, $tld, $company_id = null)
+    {
+        Loader::loadModels($this, ['Domains.DomainsTlds']);
+
+        if (is_null($company_id)) {
+            $company_id = Configure::get('Blesta.company_id');
+        }
+
+        // Fetch TLD sync settings
+        $tld_settings = $this->DomainsTlds->getDomainsCompanySettings();
+
+        $formatted_pricing = [];
+        foreach ($pricing as $currency => $terms) {
+            foreach ($terms as $year => $prices) {
+                // Apply markup and rounding
+                $prices['register'] = $this->markup(
+                    $prices['register'],
+                    $tld_settings['domains_sync_price_markup'] ?? 0,
+                    $tld_settings['domains_markup_rounding'] ?? '.00'
+                );
+                $prices['renew'] = $this->markup(
+                    $prices['renew'],
+                    $tld_settings['domains_sync_renewal_markup'] ?? 0,
+                    $tld_settings['domains_markup_rounding'] ?? '.00'
+                );
+                $prices['transfer'] = $this->markup(
+                    $prices['transfer'],
+                    $tld_settings['domains_sync_transfer_markup'] ?? 0,
+                    $tld_settings['domains_markup_rounding'] ?? '.00'
+                );
+
+                // Check if the pricing row is enabled
+                $tld_object = $this->DomainsTlds->get($tld, $company_id);
+                $pricing_row = $this->DomainsTlds->getPricing($tld_object->package_id, $year, $currency);
+
+                $formatted_pricing[$year][$currency] = [
+                    'price' => $prices['register'],
+                    'price_renews' => $prices['renew'],
+                    'price_transfer' => $prices['transfer'],
+                    'enabled' => !empty($pricing_row)
+                ];
+            }
+        }
+
+        return $formatted_pricing;
+    }
+
+    /**
+     * Applies a markup to a given price
+     *
+     * @param float $price The price to add a markup
+     * @param int $markup The percentage of markup to add
+     * @return float The total amount of the price plus the markup
+     */
+    private function markup($price, $markup, $rounding = null)
```

Comment:
I didn't have good results with this function. I had good results with this one:


```
    private function markup($price, $markup, $rounding = null)
    {
        if ($price == 0) {
            return null;
        }

        $final_price = number_format($price * (($markup / 100.00) + 1), 4, '.', '');

        if (!is_null($rounding) && is_numeric($rounding)) {
            $final_price = floor($final_price - (float) $rounding) + (float) $rounding + 1;
        }

        return $final_price;
    }
```

Reply:
Did you experience unexpected results with the proposed function?  The new one you're proposing gave me some incorrect results.  For example:
1. Round value is .2
2. Price after margin increase is 31.7424
3. Result of rounding is 31.2
4. The result should be 32.2

Reply:
Yes, when the resulting value after the margin increase was an integer, and $rounding was set to '.00', 1 was added to the final result.
i.e.: A price of $30 with a markup of 100% and rounded to the nearest zero ($rounding = '.00'), the expected result was $60, but the proposed function returned $61 instead.

Reply:
I was still having issues with the updated equation.  For example
1. Price is 10
2. Markup is 100
3. Rounding number is .10
4. Expected value is 20.10
5. Actual value is 21.10

Try this one
```
    private function markup($price, $markup, $rounding = null)
    {
        if ($price == 0) {
            return null;
        }

        $final_price = number_format($price * (($markup / 100.00) + 1), 4, '.', '');
        if (!is_null($rounding) && is_numeric($rounding)) {
            $subtracted_rounding_price = $final_price - (float) $rounding;
            $floored_price = floor($subtracted_rounding_price);
            $final_price = $floored_price + (float) $rounding + ($subtracted_rounding_price == $floored_price ? 0 : 1);
        }

        return $final_price;
    }
```